### PR TITLE
Support to use boolector as the SMT solver

### LINKED ISSUE #2403: Add boolector support
We can leverage `dev-evm-experiments` to add boolector support

https://github.com/trailofbits/manticore/blob/f0d59ce0021994b9a9b6ea3df3ccfb4719426aad/manticore/core/smtlib/solver.py#L46

https://github.com/trailofbits/manticore/blob/f0d59ce0021994b9a9b6ea3df3ccfb4719426aad/manticore/core/smtlib/solver.py#L1260-L1264

We need to add `consts.boolector_bin` to prevent `boolector` to be hardcoded.

Boolector does not work if a variable is named `BV`, and require to change
https://github.com/trailofbits/manticore/blob/3da969bf49c69c2a0f00088a96643d8b1b75264f/manticore/core/smtlib/constraints.py#L396
(master)

To

https://github.com/trailofbits/manticore/blob/f0d59ce0021994b9a9b6ea3df3ccfb4719426aad/manticore/core/smtlib/constraints.py#L397
(dev-evm-experiment)

(or to `BITVEC`, or another similar name)

### PR DESCRIPTION
**Pull request by ggrieco-tob**
This branch includes initial support for [boolector](https://github.com/Boolector/boolector). Only recent versions (e.g 3.2.x) are supported. Ancient versions like the one in Ubuntu 18.04 (1.5.x) are not. Closes #2403

### CODE REVIEW

File: manticore/core/smtlib/solver.py [hunk 1]
```diff
@@ -740,6 +742,13 @@ def __init__(self):
         super().__init__(command=command, value_fmt=10, init=init)
 
 
+class Boolector(SMTLIBSolver):
```
Comment (ekilmer) [inline manticore/core/smtlib/solver.py:745]: Nit: Would you mind changing this to follow the pattern of the solvers? `BoolectorSolver`

**Review by ekilmer — COMMENTED** [review]
This is exciting! And the process for adding a new solver looks to be very easy :) 

It would be nice to test basic functionality of `boolector` solver in CI as well.

We would need to figure out how to install it into the Ubuntu 18.04 environment and also add some tests that use `boolector`, which I think is as simple as following what's done here https://github.com/trailofbits/manticore/blob/fd69381a2c1cc7f83a1552f7d6fef762bfee1e14/tests/other/test_smtlibv2.py#L1283-L1290

Something like the following:
```python
class ExpressionTestBoolector(ExpressionTest):
    def setUp(self):
        self.solver = BoolectorSolver.instance()
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -57,18 +57,33 @@ jobs:
...
+        cd boolector
+        git checkout "f61c0dcf4a76e2f7766a6358bfb9c16ca8217224"
+        git log -1 --oneline > ../boolector.commit
+        ./contrib/setup-lingeling.sh
+        ./contrib/setup-btor2tools.sh
+        ./configure.sh
+        cd build
+        make -j4
+        mkdir -p /tmp/boolector
+        sudo make DESTDIR=/usr install
```
Comment (ekilmer) [inline .github/workflows/ci.yml:83]: Ideally, we wouldn't have to rebuild boolector each time, but that might require messing with GH Actions caching, which can be annoying.

`master` branch has this `Install dependencies` step taking around 1:30 minutes, and now it's around 2:45 or 3 minutes.

@ehennenfent any strong feelings on this? I think maybe opening an issue to figure out caching would be fine?

**Review by ekilmer — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -57,18 +57,33 @@ jobs:
...
         #install yices
         sudo add-apt-repository ppa:sri-csl/formal-methods
         sudo apt-get update
         sudo apt-get install yices2
+        #install boolector
+        mkdir -p /tmp/build
+        cd /tmp/build
+        git clone https://github.com/boolector/boolector.git
+        cd boolector
+        git checkout "f61c0dcf4a76e2f7766a6358bfb9c16ca8217224"
```
Comment (ekilmer) [inline .github/workflows/ci.yml:75]:
```suggestion
        # Version 3.2.1
        git checkout "f61c0dcf4a76e2f7766a6358bfb9c16ca8217224"
```

File: manticore/core/smtlib/solver.py [hunk 4]
```diff
@@ -282,15 +284,6 @@ def __init__(
...
             init = tuple()
         self._init = init
-        self._get_value_fmt = (
-            {
-                2: RE_GET_EXPR_VALUE_FMT_BIN,
-                10: RE_GET_EXPR_VALUE_FMT_DEC,
-                16: RE_GET_EXPR_VALUE_FMT_HEX,
-            }[value_fmt],
-            value_fmt,
-        )
```
[the remark this answers is not in the document]
Comment (ggrieco-tob) [reply manticore/core/smtlib/solver.py:292]: boolector seems to need a slightly more general parsing code, which also works for other solvers

File: manticore/ethereum/parsetab.py [hunk 5]
```diff
@@ -5,38 +5,348 @@
 
 _lr_method = "LALR"
 
-_lr_signature = "ADDRESS BOOL BYTES BYTESM COMMA FIXED FIXEDMN FUNCTION INT INTN LBRAKET LPAREN NUMBER RBRAKET RPAREN STRING UFIXED UFIXEDMN UINT UINTN\n     T : UINTN\n     T : UINT\n     T : INTN\n     T : INT\n     T : ADDRESS\n     T : BOOL\n     T : FIXEDMN\n     T : UFIXEDMN\n     T : FIXED\n     T : UFIXED\n     T : BYTESM\n     T : FUNCTION\n     T : BYTES\n     T : STRING\n\n    \n     TL : T\n    \n     TL : T COMMA TL\n    \n     T : LPAREN TL RPAREN\n    \n     T : LPAREN RPAREN\n    \n     T : T LBRAKET RBRAKET\n    \n     T : T LBRAKET NUMBER RBRAKET\n    "
+_lr_signature = "ADDRESS BOOL BYTES BYTESM COMMA FIXED FIXEDMN FUNCTION INT INTN LBRAKET LPAREN NUMBER RBRAKET RPAREN STRING UFIXED UFIXEDMN UINT UINTN\n    T : UINTN\n    T : UINT\n    T : INTN\n    T : INT\n    T : ADDRESS\n    T : BOOL\n    T : FIXEDMN\n    T : UFIXEDMN\n    T : FIXED\n    T : UFIXED\n    T : BYTESM\n    T : FUNCTION\n    T : BYTES\n    T : STRING\n\n    \n    TL : T\n    \n    TL : T COMMA TL\n    \n    T : LPAREN TL RPAREN\n    \n    T : LPAREN RPAREN\n    \n    T : T LBRAKET RBRAKET\n    \n    T : T LBRAKET NUMBER RBRAKET\n    "
```
[the remark this answers is not in the document]
Comment (ggrieco-tob) [reply manticore/ethereum/parsetab.py:8]: yeap, sorry about that :disappointed:

File: manticore/core/smtlib/solver.py [hunk 6]
```diff
@@ -351,15 +344,29 @@ def _assert(self, expression: Bool):
...
+            pattern = RE_GET_EXPR_VALUE_FMT_DEC
+            m = pattern.match(t)
+            base = 10
+        if m is None:
+            pattern = RE_GET_EXPR_VALUE_FMT_HEX
+            m = pattern.match(t)
+            base = 16
+        if m is None:
+            logger.error("I don't know how to parse the value %s from %s", str(t), expression_str)
+            assert False
```
[the remark this answers is not in the document]
Comment (ggrieco-tob) [reply manticore/core/smtlib/solver.py]: Fixed.

File: manticore/core/smtlib/solver.py [hunk 7]
```diff
@@ -351,15 +344,30 @@ def _assert(self, expression: Bool):
...
+        base = 2
+        m = pattern.match(t)
+        if m is None:
+            pattern = RE_GET_EXPR_VALUE_FMT_DEC
+            m = pattern.match(t)
+            base = 10
+        if m is None:
+            pattern = RE_GET_EXPR_VALUE_FMT_HEX
+            m = pattern.match(t)
+            base = 16
```
Comment (ekilmer) [inline manticore/core/smtlib/solver.py:358]: Minor nit, we could change to the following:
```suggestion
        base = 2
        m = RE_GET_EXPR_VALUE_FMT_BIN.match(t)
        if m is None:
            m = RE_GET_EXPR_VALUE_FMT_DEC.match(t)
            base = 10
        if m is None:
            m = RE_GET_EXPR_VALUE_FMT_HEX.match(t)
            base = 16
```

**Review by ekilmer — CHANGES_REQUESTED** [review]
Looking good!! Just a few minor things, and then I think we'll be ready for merge 🎉

**Review by ekilmer — APPROVED** [review]
LGTM. Thank you!!

**Comment by ggrieco-tob** [comment]
@ekilmer can you merge this? Github says:

> You’re not authorized to merge this pull request.

for some reason.. :thinking:

**Pull request merged**